combined_iterator.go

```diff
@@ -31,28 +33,60 @@ type combinedIterator struct {
 }
 
 type combinedIteratorConfig struct {
-	snapshotConfig snapshotIteratorConfig
-	cdcConfig      cdcIteratorConfig
+	db                    *sqlx.DB
+	tableKeys             common.TableKeys
+	fetchSize             int
+	startSnapshotPosition *common.SnapshotPosition
+	database              string
+	tables                []string
+	serverID              common.ServerID
+	mysqlConfig           *mysqldriver.Config
+	disableCanalLogging   bool
 }
 
 func newCombinedIterator(
 	ctx context.Context,
 	config combinedIteratorConfig,
 ) (common.Iterator, error) {
-	iterator := &combinedIterator{}
-	var err error
-
-	iterator.cdcIterator, err = newCdcIterator(ctx, config.cdcConfig)
+	canal, err := common.NewCanal(ctx, common.CanalConfig{
+		Config:         config.mysqlConfig,
+		Tables:         config.tables,
+		DisableLogging: config.disableCanalLogging,
+	})
 	if err != nil {
-		return nil, fmt.Errorf("failed to create cdc iterator: %w", err)
+		return nil, fmt.Errorf("failed to start canal at combined iterator: %w", err)
 	}
 
-	iterator.snapshotIterator, err = newSnapshotIterator(ctx, config.snapshotConfig)
+	snapshotIterator, err := newSnapshotIterator(ctx, snapshotIteratorConfig{
+		getMasterPos:  canal.GetMasterPos,
+		db:            config.db,
+		tableKeys:     config.tableKeys,
+		fetchSize:     config.fetchSize,
+		startPosition: config.startSnapshotPosition,
+		database:      config.database,
+		tables:        config.tables,
+		serverID:      config.serverID,
+	})
 	if err != nil {
 		return nil, fmt.Errorf("failed to create snapshot iterator: %w", err)
 	}
 
-	iterator.currentIterator = iterator.snapshotIterator
+	cdcIterator, err := newCdcIterator(ctx, cdcIteratorConfig{
+		canal:       canal,
+		tables:      config.tables,
+		mysqlConfig: config.mysqlConfig,
+		position:    &snapshotIterator.cdcStartPosition,
```

Comment:
It may be easier, to keep the master position acquisition as part of the cdcIterator, since it is relevant to that iterator exclusively. 

```
unlock, err := cdcIterator.MasterPos(ctx)
defer unlock()

snapshotIterator, err := newSnapshotIterator(...) // retrieve min/max in values. 
```